### ISSUE TITLE
Integrate clipboard.js for docs

### ISSUE THREAD

**Issue by coliff** [squidfunk/mkdocs-material#315]
Feature request: Integrate clipboard.js for docs - maybe as an optional plugin?
https://clipboardjs.com/

**Comment by squidfunk** [comment]
Are you speaking of integration for the code sections or what in particular?

**Comment by coliff** [comment]
Yes, for code sections (sorry I didn't make that clearer)

**Comment by facelessuser** [comment]
This seems like something that could be done with some custom JavaScript without requiring Material to include this.  I feel like you could just have some JS that targets the code blocks, adds a unique id, and then injects a clipboard button linked to the that unique id.

Having it built-in is fine though if that is the direction that is wanted, but I feel like you could get that functionality now without Material having to intervene.  That's just the impression I get from reading `cliboard.js`'s documentation.  I haven't tried it yet.

**Comment by facelessuser** [comment]
This is just a real quick proof of concept.  This is hunting for code highlighted under `<div class="codehilite"><pre>`.

Some CSS:

```css
.md-typeset div.codehilite {
  position: relative;
  overflow: visible;
  padding: 0;
}
.md-typeset div.codehilite pre{
  overflow-x: auto;
  display: block;
  padding: 1rem 1.2rem .8rem;
}
.md-typeset .btn {
  float: right;
  display: inline-block;
  position: absolute;
  top: 3px;
  right: 3px;
  opacity: 0;
  transition: opacity 0.3s ease-in-out;
  background-color: #eee;
  background-image: linear-gradient(#fcfcfc,#eee);
  border: 1px solid #d5d5d5;
  border-radius: 3px;
  padding: 6px 12px;
  cursor: pointer;
}
.md-typeset .btn:hover {
  opacity: 1;
  transition: opacity 0.3s ease-in-out;
}
```

Some JavaScript to convert these code blocks into clipboard ready blocks. `doc_theme/extra.js`.

```js
(function (document) {
    function clipboard_init() {
        var charts = document.querySelectorAll("div.codehilite"),
            arr = [],
            i, j, maxItem, pre, btn, el;

        // Make sure we are dealing with an array
        for(i = 0, maxItem = charts.length; i < maxItem; i++) arr.push(charts[i]);

        // Find the UML source element and get the text
        for (i = 0, maxItem = arr.length; i < maxItem; i++) {
            el = arr[i];
            pre = el.childNodes[0];

            pre.id = "hl_code" + i.toString();
            btn = document.createElement('button');
            btn.appendChild(document.createTextNode('copy'));
            btn.setAttribute("class", "btn");
            btn.setAttribute("data-clipboard-target", "#hl_code" + i.toString());
            el.insertBefore(btn, pre);
        }
        new Clipboard('.btn');
    };

    function onReady(fn) {
        if (document.addEventListener) {
            document.addEventListener('DOMContentLoaded', fn);
        } else {
            document.attachEvent('onreadystatechange', function() {
                if (document.readyState === 'interactive')
                    fn();
            });
        }
    }

    onReady(function(){
        clipboard_init();
    });
})(document);
```

Add following to `mkdocs.yml`.

```yml
theme_dir: doc_theme
extra_javascript:
  - https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.6.0/clipboard.min.js
  - extra.js
```

![clip](https://cloud.githubusercontent.com/assets/1055125/25830699/ccbf7c78-341b-11e7-95b7-10934703a462.gif)

**Comment by Timber232** [comment]
Thanks for the example @facelessuser! 

I got mine to work but I had to change my `mkdocs.yml` to the following:

```
extra_css:
  - doc_theme/extra.css
extra_javascript:
  - https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.6.0/clipboard.min.js
  - doc_theme/extra.js
markdown_extensions:
  - codehilite
```

For your example to work you need `codehilite` extension and explicitly specify which folder the extra.js and extra.css is located (in this case `doc_theme`)

**Comment by facelessuser** [comment]
> For your example to work you need codehilite extension as bare minimum and explicitly specify which folder the extra.js and extra.css is located (in this case doc_theme).

Yeah, I was making some assumptions, but yes, `codehilite` gets you the `<div><pre>` blocks, but you could apply this to traditional `<pre><code>` as well, you would need to adjust accordingly.

> I would prefer this to be implemented in the MkDocs Material theme though. I know we can change the CSS ourselves, but it would be nice to have the "Copy" button fully integrated (using Material button, animation and all) with the MkDocs Material theme out of the box.

Understood.  With a little CSS you can get the feel of material, but I understand the idea of having it working out of the box.  I'm not sure how much @squidfunk plans on supporting out of the box, he may or may not want to bother with clipboard.js.  So I just wanted to show it is fairly easy to get there.  To be honest, I was kind of meh on about this, but after throwing it together, I'm kind of liking it not I got it more material-ish 🙂 .

![screenshot 2017-05-08 20 36 54](https://cloud.githubusercontent.com/assets/1055125/25833390/397c5c3e-342e-11e7-971e-9ad887362fc5.png)

**Comment by coliff** [comment]
@facelessuser Thanks so much for the integration example. My JavaScript skills aren't so great and I didn't know how to add that myself. I copied your code examples and got it working great on my site. One suggestion for improvement though would be for the 'copy' button to appear when you hover over any part of the code block, at the moment it only showed when you hovered over the top right, so if you didn't know it was there then you'd likely miss it.

I added this additional CSS so it appears when hovering over the code block;

```
.codehilite:hover .btn {
  opacity: 0.9;
  transition: opacity .3s ease-in-out;
}
```

**Comment by squidfunk** [comment]
@facelessuser this looks great, thanks for prototyping it. Will integrate this somehow in the theme, but I agree with @coliff that it should only be visible on hover.

**Comment by facelessuser** [comment]
> Will integrate this somehow in the theme, but I agree with @coliff that it should only be visible on hover.

Come on guys, I wasn't finished yet 🙂 . I went ahead and pushed what I had finished last night: 
 https://github.com/facelessuser/pymdown-extensions/commit/3036f6790d7201ac9859240b4ca994df45522fa2.  You can see it in action here: http://facelessuser.github.io/pymdown-extensions/extensions/superfences/.

- When hovering over the code div, a muted icon will appear in the upper right to show it is there.
- When mousing over the muted icon, it will then be fully visible.
- When clicking copy, it will text will get copied to the clipboard and the text selection will be cleared up.
- Lastly a little tooltip will appear letting you know the copy was a success, or if there was an error.  I pop it up to the right because when using line numbers, the code block is in a table and the right edge cuts the tooltip off.

I'm not a web developer, so I look forward to an official solution that is much more polished, but hopefully if anyone is itching to get something rolling, maybe this will be helpful until an official solution is available.

**Comment by squidfunk** [comment]
Feel free to submit a PR. I will review it and may make some changes. Also some cross browser testing (+ mobile devices) is necessary.

**Comment by facelessuser** [comment]
> Feel free to submit a PR. I will review it and may make some changes. Also some cross browser testing (+ mobile devices) is necessary.

I'll see what I can do.  I need to get familiar with setting of the environment and building it.  So much frontend web framework stuff 😮 .

**Comment by squidfunk** [comment]
Okay, no worries, I can also review and incorporate the stuff from your commit.

**Comment by facelessuser** [comment]
I will definitely give it a go.  If you don't hear from me by the end of the week, the frontend framework defeated me 😄 .

**Comment by facelessuser** [comment]
So far it does seem to work on Chrome, Firefox, and Safari on Mac. I was more worried about mobile, but it works great on Android. Don't think I have an iOS device laying around though.

**Comment by facelessuser** [comment]
Had a much easier time with the dev environment this time.  I can figure out everything else from here, so I will issue a pull when done.

**Comment by squidfunk** [comment]
Okay, great!

**Comment by facelessuser** [comment]
@squidfunk, Branch is [here](https://github.com/facelessuser/mkdocs-material/tree/clipboardjs), but we probably need to talk about how exactly clipboard.js is to be integrated.  Currently, I have it so the user adds clipboard.js to `extra_javascript` and  have a check to see if `Clipboard` is `undefined` before injecting the clipboard overlays, but I suspect this is not going to be what is desired.  Anyways, before I do a pull, I just want to get some direction on what is desired.

**Comment by squidfunk** [comment]
I will review it as soon as possible!

**Comment by facelessuser** [comment]
So it seems to work generally on IE and Edge.  The notification tooltip doesn't work on IE though.  It's not crucial, but I think it should work; it's nice to have something let you know the copy succeeded.  So I'd have to figure out a tooltip that would work on IE and others.  This might be where a front end expert may have to fill in my gaps in knowledge.

**Comment by makshh** [comment]
@facelessuser I think the issue with tooltip not being displayed in IE is `overflow` property in CSS. It seems to work with `overflow: visible` on `.clip-btn`

**Comment by facelessuser** [comment]
@makshh, you sir, are a genius :clap:.  That is exactly what I was missing.  And I would have been chasing my tail trying to figure that one out.

I'll commit that later today and try to rebase and squash all of this down to a single commit.

**Comment by facelessuser** [comment]
IE fix is in and I squashed all the commits in this new branch here: https://github.com/facelessuser/mkdocs-material/commits/clipboard.

**Comment by facelessuser** [comment]
Went ahead and issued a pull where we can discuss the changes and get this polished up.

**Comment by squidfunk** [comment]
Closing this, as it is implemented in PR #333, which is currently being verified. When the PR is merged, I will issue a new release.
